react/AppIcon/index.jsx

```diff
@@ -7,16 +7,129 @@ import Icon from '../Icon'
 
 import appDefaultIcon from '../../assets/icons/ui/cube.svg'
 
+let domain
+const _getDomain = () => {
+  if (domain) return domain
+  const root = document.querySelector('[role=application]')
+  if (!root) {
+    throw new Error('Cannot find [role=application] element')
+  }
+  domain = root.dataset.cozyDomain
+  if (!domain) {
+    throw new Error(
+      'Application element does not contain `data-cozy-domain` attribute'
+    )
+  }
+  return domain
+}
+
+let protocol
+const _getProtocol = () => {
+  if (protocol) return protocol
+  protocol = window.location.protocol
+  return protocol
+}
+
+const _preload = url => {
+  return new Promise((resolve, reject) => {
+    const loader = document.createElement('img')
+    loader.onload = () => {
+      resolve(url)
+    }
+    loader.onerror = () => {
+      reject(new Error(`Error while preloading ${url}`))
+    }
+    loader.src = url
+  })
+}
+
+const _getAppIconURL = app => {
+  let path = _getInstalledIconPath(app)
+  path = path || _getRegistryIconPath(app)
+  return path ? `${_getProtocol()}//${_getDomain()}${path}` : null
+}
+
+const _getInstalledIconPath = app => {
+  return app.links && app.links.icon
+}
+
+const _getRegistryIconPath = app => {
+  return (
+    app.latest_version &&
+    app.latest_version.version &&
+    `/registry/${app.slug}/${app.latest_version.version}/icon`
+  )
+}
+
+const _loaded = []
+
+const DONE = 'done'
+const ERRORED = 'errored'
+const FETCHING = 'fetching'
+const IDLE = 'idle'
+
 export class AppIcon extends Component {
   constructor(props, context) {
     super(props, context)
+
+    const { app } = props
+    let source
+    try {
+      source = _getAppIconURL(app)
+    } catch (error) {} // eslint-disable-line no-empty
+
+    const loaded = source && _loaded.includes(source)
+
     this.state = {
-      status: 'idle'
+      icon: loaded ? source : null,
+      status: loaded ? DONE : IDLE
     }
   }
 
   async componentDidMount() {
-    this.fetchIcon()
+    const { icon } = this.state
+    if (icon) return
+
+    try {
+      await this.load()
+    } catch (error) {
+      const { fetchIcon } = this.props
+      if (fetchIcon) {
+        await this.fetchIcon()
+      } else {
+        console.error(error.message)
+        this.setState({ status: ERRORED, error: error })
+      }
+    }
+  }
+
+  async load() {
+    this.setState({ status: FETCHING })
```

Comment:
The FETCHING/DONE state is managed at two places in this class, it would be clearer it was managed only at one place.

Reply:
The thing is, it's managed at the beginning when the component starts to load the icon, and updated again when it's done. The thing is ti should be done directly in the `componentDidMount` method I think.